parser-typechecker/src/Unison/Util/Rope.hs

```diff
@@ -0,0 +1,197 @@
+{-# Language FunctionalDependencies #-}
+{-# Language DeriveFoldable #-}
+
+module Unison.Util.Rope 
+  (chunks, singleton, one, map, traverse, null, flatten, two, cons, uncons, snoc, unsnoc, 
+   Sized(..), Take(..), Drop(..), Reverse(..), Index(..), Rope,
+   ) 
+where
+
+import Prelude hiding (drop,take,reverse,map,traverse,null)
+import Data.Foldable (toList)
+import Control.DeepSeq (NFData(..))
+
+-- | Size-balanced binary tree of chunks. 
+-- The `Int` in the `Two` constructor is a cached size of that subtree.
+data Rope a
+  = Empty
+  | One {-# unpack #-} !a
```

Comment:
Nit: unpack doesn't work on polymorphic fields

Reply:
The compiler accepts it if the field is strict (but not if it's lazy), which made me think a `newtype Text = Text (Rope Data.Text)` where the `a` is now specialized could unpack the `Data.Text` into the constructor. Do you know for sure that won't happen?

Or is there some other way to tell Haskell to produce specialized versions of `Rope` for particular instantiations of `a`?

cc @dolio